### PR TITLE
ns for fx: support comparing fp32 vs fp32_prepared/int8, except shadowed

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61380
* #61377
* #61360
* #61359
* #61358
* #61357
* #61323
* **#61129**

Summary:

Adds support the comparing fp32 model (without quantization) to a
fp32 model prepared with quantization or a quantized model. The main missing feature was
handling conv-bn fusion, since this fusion for PTQ happens outside
of quantization patterns.

Adds testing for this case for comparing weights and comparing
activations

Adds a TODO for also handling this for shadow activations, we need to
first stop removing observers in graph passes before we can add
this support, will be in a future PR.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2
python test/test_quantization.py TestFXGraphMatcherModels.test_mobilenet_v2_qat
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels.test_compare_activations_conv
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29520009](https://our.internmc.facebook.com/intern/diff/D29520009)

### CODE REVIEW

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29520009).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29520009).

File: test/quantization/fx/test_numeric_suite_fx.py [hunk 1]
```diff
@@ -758,19 +762,21 @@ def _test_extract_weights(
...
         # test both the public API as well as the internal GraphModule API
         for extract_weights_fun in (extract_weights, _extract_weights_impl):
-            results = extract_weights_fun('fp32_prepared', mp, 'int8', mq)
-            self.assertTrue(
-                len(results) == results_len,
-                f"expected len {results_len}, got len {len(results)}")
-            self.assert_ns_compare_dict_valid(results)
+            # test both m vs mp and mp vs mq
+            for m1, m2 in ((m, mp), (mp, mq)):
+                results = extract_weights_fun('a', mp, 'b', mq)
```
Comment (raghuramank100) [inline test/quantization/fx/test_numeric_suite_fx.py:775]: Should this be m1 and m2?

**Review by raghuramank100 — APPROVED** [review]

File: test/quantization/fx/test_numeric_suite_fx.py (hunk 1 above)
Comment (vkuzo) [reply test/quantization/fx/test_numeric_suite_fx.py:775]: yes, great catch, will fix that

Comment (vkuzo) [reply test/quantization/fx/test_numeric_suite_fx.py:775]: yes, great catch, will fix that

**Review by hx89 — APPROVED** [review]
lgtm! Just wonder what would be the use case for comparing fp32 and fp32_prepared? Is it for sanity check to make sure fp32 is numerically the same after prepare and helping debug the implementation of prepare function?

**Comment by vkuzo** [comment]
> lgtm! Just wonder what would be the use case for comparing fp32 and fp32_prepared? Is it for sanity check to make sure fp32 is numerically the same after prepare and helping debug the implementation of prepare function?

there was a use case discussed during design review to compare fp32 to quantized.  This PR implements this functionality (once you can compare A to B and B to C, you can compare A to C).  Let me update the summary to say this.

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29520009).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29520009).

**Pull request closed without merging**